src/aiori-NC4.c

```diff
@@ -0,0 +1,355 @@
+
+static char *GetNamedMessage(char * message, char * name);
+
+static void *NC4_Create(char *, IOR_param_t *);
+
+static void *NC4_Open(char *, IOR_param_t *);
+
+static IOR_offset_t NC4_Xfer(int, void *, IOR_size_t *,
+                             IOR_offset_t, IOR_param_t *);
+
+static void NC4_Close(void *, IOR_param_t *);
+
+static void NC4_Delete(char *, IOR_param_t *);
+
+static char *NC4_GetVersion();
+
+static void NC4_Fsync(void *, IOR_param_t *);
+
+static IOR_offset_t NC4_GetFileSize(IOR_param_t *, MPI_Comm, char *);
+
+static int NC4_Access(const char *, int, IOR_param_t *);
+
+/************************** D E C L A R A T I O N S ***************************/
+
+ior_aiori_t nc4_aiori = {
+        .name = "NC4",
+        .name_legacy = NULL,
+        .create = NC4_Create,
+        .open = NC4_Open,
+        .xfer = NC4_Xfer,
+        .close = NC4_Close,
+        .delete = NC4_Delete,
+        .get_version = NC4_GetVersion,
+        .fsync = NC4_Fsync,
+        .get_file_size = NC4_GetFileSize,
+        .statfs = aiori_posix_statfs,
+        .mkdir = aiori_posix_mkdir,
+        .rmdir = aiori_posix_rmdir,
+        .access = NC4_Access,
+        .stat = aiori_posix_stat,
+};
+
+/***************************** F U N C T I O N S ******************************/
+
+/*
+ * Create and open a file through the NC4 interface.
+ */
+static void *NC4_Create(char *testFileName, IOR_param_t *param) {
+    int *fd;
+    int fd_mode;
+    MPI_Info mpiHints = MPI_INFO_NULL;
+
+    SetHints(&mpiHints, param->hintsFileName);
+    if (rank == 0 && param->showHints) {
+        fprintf(stdout, "\nhints passed to nc_create_par() {\n");
+        ShowHints(&mpiHints);
+        fprintf(stdout, "}\n");
+    }
+
+    fd = (int *) malloc(sizeof(int));
+    if (fd == NULL)
+        ERR("malloc() failed");
+
+    fd_mode = GetFileMode(param);
+    NC4_CHECK(nc_create_par(testFileName, fd_mode, testComm, mpiHints, fd),
+              "cannot create file");
+
+    return (fd);
+}
+
+/*
+ * Open a file through the NC4 interface.
+ */
+static void *NC4_Open(char *testFileName, IOR_param_t *param) {
+    int *fd;
+    int fd_mode;
+    MPI_Info mpiHints = MPI_INFO_NULL;
+
+    SetHints(&mpiHints, param->hintsFileName);
+    if (rank == 0 && param->showHints) {
+        fprintf(stdout, "\nhints passed to nc_open_par() {\n");
+        ShowHints(&mpiHints);
+        fprintf(stdout, "}\n");
+    }
+
+    fd = (int *) malloc(sizeof(int));
+    if (fd == NULL)
+        ERR("malloc() failed");
+
+    fd_mode = GetFileMode(param);
+    NC4_CHECK(nc_open_par(testFileName, fd_mode, testComm, mpiHints, fd),
+              "cannot open file");
+
+    return (fd);
+}
+
+/*
+ * Write or read access to file using the NC4 interface.
+ */
+static IOR_offset_t NC4_Xfer(int access, void *fd, IOR_size_t *buffer,
+                             IOR_offset_t length, IOR_param_t *param) {
+    signed char *bufferPtr = (signed char *) buffer;
+    static int firstReadCheck = FALSE, startDataSet;
```

Comment:
The use of a static variable here appears problematic. As IOR could repeat an experiment multiple times, this shouldn't be a static variable.

Reply:
Thanks, @JulianKunkel!  I'll take a closer look at the logic here.  To be honest, I cribbed this code from the NCMPI aiori, which also uses a `static` declaration.  See:

https://github.com/hpc/ior/blob/d49d345007530a91e9f36dfdab87bb46c006a674/src/aiori-NCMPI.c#L176-L180

I think that Wei-keng Liao had intended to use this static variable to only enable certain code for the first call of `*_Xfer` on a `READCHECK`.  So, I think that logic requires it to be static.

However, I'm happy to rethink the logic here and implement a different solution, if you'd prefer.